rclcpp/include/rclcpp/executor.hpp

```diff
@@ -319,6 +320,53 @@ class Executor
   virtual void
   spin_once(std::chrono::nanoseconds timeout = std::chrono::nanoseconds(-1));
 
+  // TODO(hliberacki): Add documentation
+  template<typename FutureT, typename DurationT = std::chrono::milliseconds,
+    // typename std::enable_if_t<
+    //   std::is_same_v<
+    //     std::remove_reference_t<FutureT>,
+    //     std::future<
+    //       decltype(std::declval<std::remove_reference_t<FutureT>>()
+    //       .get())>> ||
+    //   std::is_same_v<
+    //      std::remove_reference_t<FutureT>,
+    //      std::shared_future<
+    //        decltype(std::declval<std::remove_reference_t<FutureT>>()
+    //        .get())>>,
+    //   bool> = true>
+    typename std::enable_if_t<!std::is_invocable_v<FutureT>, bool> = true>
+  FutureReturnCode spin_until_complete(FutureT & future, DurationT timeout = DurationT(-1))
+  {
+    // call helper wait untill complete with lambda
+    auto checkFuture = [&future]() {
+        return future.wait_for(std::chrono::seconds(0)) ==
+               std::future_status::ready;
+      };
+    return spin_until_complete_impl(checkFuture, timeout);
+  }
+
+  template<
+    typename Condition, typename DurationT = std::chrono::milliseconds,
+    typename std::enable_if_t<std::is_invocable_v<Condition>, bool> = true>
+  FutureReturnCode spin_until_complete(
+    Condition & condition,
+    DurationT timeout = DurationT(-1))
+  {
+    using RetT = decltype(std::declval<std::remove_reference_t<Condition>>()());
+    static_assert(
+      std::is_same_v<bool, RetT>,
+      "Conditional callable has to return boolean type");
+    return spin_until_complete_impl(condition, timeout);
+  }
+  // TODO(hliberacki): Add documentation
+  /// spin (blocking) for given amount of time
+  template<typename DurationT>
+  void
+  spin_for(DurationT timeout)
+  {
+    return spin_until_complete([]() {return false;}, timeout);
```

Comment:
I guess the compiler can be clever and optimize the implementation well enough, but maybe having a direct implementation is better.

Reply:
@ivanpauno Sorry maybe it's the hour (past midnight in my time-zone ;) ) But I don't get that comment. I'd like to reimplement the logic here as well? I thought it might be a waste of time since it's exactly the same logic the difference is only calling this lambda which always returns true - so as you've said probably most compilers will just create a constant value there. 

But sure I can just provide direct implementation here as well if needed. Should not take more than a few minutes.